src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -3078,6 +3086,72 @@ private BoundExpression BindArgumentExpression(BindingDiagnosticBag diagnostics,
+                {
+                    if (lazyInferredOutArgumentSafeToEscape is { } inferredOutArgumentSafeToEscape)
+                    {
+                        return inferredOutArgumentSafeToEscape;
+                    }
+
+                    // the safe-to-escape of any out parameters in the method we are calling.
+                    uint outParamSafeToEscape = UseUpdatedEscapeRulesForInvocation(leastOverriddenMember) ? ReturnOnlyScope : CallingMethodScope;
+
+                    // the narrowest safe-to-escape of any argument which can escape (either by-value or by-reference) into the argument for any out parameter.
+                    uint narrowestArgSafeToEscape = CallingMethodScope;
+
+                    // TODO: try and reuse arg-mixing bits to reduce redundant logic here
+                    for (var arg = 0; arg < arguments.Count; arg++)
+                    {
+                        var paramNum = result.ParameterFromArgument(arg);
+                        var sourceParam = paramNum < parameters.Length ? parameters[paramNum] : null;
+                        if (sourceParam is null)
+                        {
+                            var arglist = (BoundArgListOperator)arguments[arg];
+
+                            // arglist arguments can never escape by-ref into out parameters
+                            const uint ArglistRefSafeToEscape = CurrentMethodScope;
+                            Debug.Assert(ArglistRefSafeToEscape > outParamSafeToEscape);
+
+                            // arglist arguments can always escape by-value into out parameters
+                            const uint ArglistValSafeToEscape = CallingMethodScope;
+                            Debug.Assert(ArglistValSafeToEscape <= outParamSafeToEscape);
+
+                            for (var arglistArg = 0; arglistArg < arglist.Arguments.Length; arglistArg++)
+                            {
+                                var refKind = arglist.ArgumentRefKindsOpt.IsDefault || arglist.ArgumentRefKindsOpt.Length < arglistArg ? RefKind.None : arglist.ArgumentRefKindsOpt[arglistArg];
+                                narrowestArgSafeToEscape = Math.Max(narrowestArgSafeToEscape, GetValEscape(arglist.Arguments[arglistArg], LocalScopeDepth));
+                            }
+
+                            continue;
+                        }
+
+                        findNarrowestArgumentEscape(sourceParam, arguments[arg]);
+                    }
+
+                    if (receiver is not null && leastOverriddenMember.EnclosingThisSymbol() is { } thisParam)
+                    {
+                        findNarrowestArgumentEscape(thisParam, receiver);
+                    }
+
+                    lazyInferredOutArgumentSafeToEscape = narrowestArgSafeToEscape;
+                    return narrowestArgSafeToEscape;
+
+                    void findNarrowestArgumentEscape(ParameterSymbol sourceParam, BoundExpression argument)
```

Comment:
This is re-implementing the logic it took us so long to agree on in my previous PR. Duplicating the logic at a different layer in a different way seems like it's not the right path forward.

Reply:
I agree, I just don't know what that factoring looks like yet.